whitepapers/0010_ibc_generic_messaging

```diff
@@ -0,0 +1,92 @@
+
+## Detailed Design
+
+### External Chain -> Cosmos Chain
+
+This will work exactly the same way it works today. We will deploy our wormhole cosmwasm contract stack to every cosmos chain we want to support. Wormhole relayers will be able to post VAAs produced for any source chain directly to the cosmos destination chain.
+
+### Cosmos Chain -> External Chain
+
+Typically, the wormhole core bridge contract emits a message which the guardians then pick up from their full nodes.
+
+For cosmos chains, we update the core bridge contract to instead send this message over IBC to wormchain. Then a wormchain contract receives the message to emit and actually emits it, which the guardians then pick up.
+
+Specifically, we implement two new cosmwasm smart contracts: `wormhole-ibc` and `wormchain-ibc-receiver`.
+
+The `wormhole-ibc` contract is meant to replace the `wormhole` core bridge contract on cosmos chains. It imports the `wormhole` contract as a library and delegates core functionality to it before and after running custom logic:
+- During execution, it delegates all logic to the core bridge library. After the core bridge library has finished execution, it will send the response as an IBC message to the `wormchain-ibc-receiver` contract <b>if and only if</b> the execution message is of type `wormhole::msg::ExecuteMsg::PostMessage`.
+
+Sending an IBC packet requires choosing an IBC channel to send over. IBC `(channel_id, port_id)` pairs are unique. All IBC-enabled cosmwasm contracts follow a standard `port_id` format: `wasm.<contract_address>`. Therefore, to choose a correct channel the `wormhole-ibc` contract will need to know the address of the `wormchain-ibc-receiver` contract - it will select any channel that is paired with the expected port. 
+
+The `wormchain-ibc-receiver` contract will be deployed on wormchain and is meant to receive the IBC messages that the `wormhole-ibc` contract sends from various cosmos chains. Its only responsibility is to receive the IBC message, send an IBC acknowledgement to the source chain, and then emit the message for the guardian node to observe.
+
+### IBC Relayers
+
+All IBC communication is facilitated by [IBC relayers](https://ibcprotocol.org/relayers/). Since these are lightweight processes that need to only listen to blockchain RPC nodes, each (only several is also acceptable) wormhole guardian can run a relayer.
+
+The guardian IBC relayers are configured to connect the `wormchain-ibc-receiver` contract on wormchain to the various `wormhole-ibc` contracts on the cosmos chains that wormhole supports.
+
+### Guardian Node Watcher
+
+We will modify the cosmos guardian watcher to watch the `wormchain-ibc-receiver` contract on wormchain for the messages from the designated `wormhole-ibc` contracts on supported cosmos chains. This is nearly identical to the current model and can be a drop in replacement.
```

Comment:
```suggestion
We will modify the ibc enabled guardian watcher to watch the `wormchain-ibc-receiver` contract on wormchain for the messages from the designated `wormhole-ibc` contracts on supported ibc enabled chains. This is nearly identical to the current model and can be a drop in replacement.
```

There is a [cosmwasm](https://github.com/wormhole-foundation/wormhole/tree/a846036b6ebff3af6f12ff375f5c3801ada20291/node/pkg/watchers/cosmwasm) and a [wormchain](https://github.com/wormhole-foundation/wormhole/tree/a846036b6ebff3af6f12ff375f5c3801ada20291/node/pkg/watchers/wormchain) watcher, but there is no cosmos watcher as wormhole doesn't (yet) support cosmos and it's not a lib either.